firebase-sessions/src/main/kotlin/com/google/firebase/sessions/api/FirebaseSessionsDependencies.kt

```diff
@@ -0,0 +1,91 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.sessions.api
+
+import androidx.annotation.VisibleForTesting
+import java.util.Collections.synchronizedMap
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+/**
+ * [FirebaseSessionsDependencies] determines when a dependent SDK is installed in the app. The
+ * Sessions SDK uses this to figure out which dependencies to wait for to getting the data
+ * collection state. This is thread safe.
+ *
+ * This is important because the Sessions SDK starts up before dependent SDKs.
+ */
+object FirebaseSessionsDependencies {
+  private val dependencies = synchronizedMap(mutableMapOf<SessionSubscriber.Name, Dependency>())
+
+  /**
+   * Add a subscriber as a dependency to the Sessions SDK. Every dependency must register itself, or
+   * the Sessions SDK will never generate a session.
+   */
+  fun addDependency(subscriberName: SessionSubscriber.Name) {
+    if (dependencies.containsKey(subscriberName)) {
+      throw IllegalArgumentException("Dependency $subscriberName already added.")
+    }
+
+    // The dependency is locked until the subscriber registers itself.
+    dependencies[subscriberName] = Dependency(Mutex(locked = true))
+  }
+
+  /** Register and unlock the subscriber. This must be called before [getSubscribers] can return. */
+  internal fun register(subscriber: SessionSubscriber) {
+    val subscriberName = subscriber.sessionSubscriberName
+    val dependency = getDependency(subscriberName)
+
+    dependency.subscriber?.run {
+      throw IllegalArgumentException("Subscriber $subscriberName already registered.")
+    }
+    dependency.subscriber = subscriber
+
+    // Unlock to show the subscriber has been registered, it is possible to get it now.
+    dependency.mutex.unlock()
+  }
+
+  /** Gets the subscribers safely, blocks until all the subscribers are registered. */
+  internal suspend fun getSubscribers(): Map<SessionSubscriber.Name, SessionSubscriber> {
+    // The call to get will never throw because the mutex guarantees it's been registered.
+    return dependencies.mapValues { (subscriberName, dependency) ->
+      dependency.mutex.withLock { get(subscriberName) }
+    }
+  }
+
+  /** Gets the subscriber regardless of being registered. This is exposed for testing. */
+  @VisibleForTesting
+  internal operator fun get(subscriberName: SessionSubscriber.Name): SessionSubscriber {
+    return getDependency(subscriberName).subscriber
+      ?: throw IllegalStateException("Subscriber $subscriberName has not been registered.")
+  }
+
+  /** Resets the registered subscribers for testing purposes. */
+  @VisibleForTesting
+  internal fun reset() {
```

Comment:
Should we clean up the dependency list here as well? `reset` makes me feel that all the dependencies/registrations will be cleaned up.

Reply:
I changed this to clear all depedencies, and updated the tests. This is cleaner.